ls1intum/Artemis
# [Bugfix] Fix Delete action for Exams/ExerciseGroups/Exercises

### Checklist

### Motivation and Context
This PR is a follow-up to #1989 aimed at improving the delete functionality of exercise groups and exams. 
Deleting exams or exam elements does not work reliably. There have been several issues identified: 
Deleting exercises from exams is not possible, they are not removed. 
Ensure that all elements are deleted correctly (participations/build plans/repositories) etc. 
Fixes #1975
Fixes #1976

**Edit:** Additionally since it fits the functionality, I also refactored the delete function for courses in this PR. Firstly i moved the delete function to the service (similar to exercises and exams), and use the new `examService.deleteById()` to ensure that the exams are properly deleted. I also improved the documentation and enhanced the delete integration test.

### Description
- [x] Refactor delete exam & delete course, move to service
- [x] Fix #1975 by cascading the remove operation of text exercises. Extend integration test to ensure this.
- [x] #1976 could not be reproduced. We added an integration test to verify this and quiz exercises with short answer questions are deleted correctly. Therefore this issue can also be closed.
- [x] Improve warning message when the instructor is about to delete an exam or course. We now warn him that student exams, buildplans and repositories will also be deleted. 
- [x] If an exercise is now deleted which is part of existing student exams, we just remove the reference from the student exam. We do not delete the student exam. 
- [x] If you delete an exam, all buildplans & repositories are deleted.
- [x] If you want to delete an exercise group which contains a programming exercise, you are presented with the choice to also delete the buildplans & repositories. 
- [x] Create Integration test: `testDeleteExamWithStudentExamsAfterConductionAndEvaluation`. This exam contains all exercise types including all different quiz type questions (#1976).
- [x] Fix flaky StudentExam integration tests
- [x] Improve `testDeleteCourseWithPermission`
- [x] Improve Documentation

### Steps for Testing
_Delete Exam General_
1. Log in to Artemis
2. Navigate to Course Administration/Exams
3. Try deleting exercise groups with different exercises. 
4. Delete old exams and cleanup your Test Course :)
5. In case of programming exercises, double-check that buildplans & repositories are deleted as well
-----------------------------------------------------------------------------------------------
_Student Exams on delete exercise_
1. Create an exam with multiple exercises and generate the student exams
2. Delete an exercise or exercise group. 
3. Check that the student exams are not deleted and that they do not reference the deleted exercise anymore(It shouldn't show up).
------------------------------------------------------------------------------------------------
_Delete Course_
1. Delete one of the test courses, ideally, it should contain as many elements as possible. 
2. Everything should work as expected.
